### PR TITLE
Add caching system and example

### PR DESCRIPTION
**Pull request by AlexandreAbraham**

### CODE REVIEW

File: examples/plot_replay_cache.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+test_run()
+list_files('./cache')
+
+#############################################################################
+# Being a bit paranoid, we would like to check what cardinal does. For that,
+# we compute the batch size of each iteration. Fortunately, we have cached
+# the variable `selected` and therefore, we can replay the experiment.
+
+
+def calc_batch_size(previous, current):
```
Comment (JoachimZ) [inline examples/plot_replay_cache.py:127]: It's a bit complex and hard to read, it's only for paranoid people, not for a base example, and it adds a bit of complexity to the main function with the compute_metric thing. So I would trash all that.

**Review by JoachimZ — CHANGES_REQUESTED** [review]
Cool! I only looked at the client code for now (example)

File: examples/plot_replay_experiment.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+            index = GrowingIndex(X_train.shape[0])
+
+            # Add at least one sample from each class
+            index.add_to_selected([np.where(y_train == i)[0][0] for i in np.unique(y)])
+
+            selected = cache.variable('selected', index.selected)
+
+            for j, prev_selected in cache.iter(range(n_iter), selected.previous()):
+                print('Computing iteration {}'.format(j))
+                index.resume(prev_selected)
```
Comment (dsleo) [inline examples/plot_replay_experiment.py:92]: I've only read the client code here but it's not clear or at least not easy to guess what this does ?

File: examples/plot_replay_experiment.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+                model.fit(X_train[prev_selected], y_train[prev_selected])
+                sampler.fit(X_train[prev_selected], y_train[prev_selected])
+                index.add_to_selected(sampler.select_samples(X_train[index.non_selected]))
+                selected.set(index.selected)
```
Comment (dsleo) [inline examples/plot_replay_experiment.py:97]: so `set` here is to be understand as the verb, not the noun ?

File: examples/plot_replay_experiment.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+such as selected samples, in a cache. Users can therefore resume
+an interrupted experiment, using the ResumeCache, or replay an entire
+experiment to perform additional computation such as metric using the
+ReplayCache.
+"""
+
+import shutil
+import os
+import numpy as np
+import dataset
```
Comment (dsleo) [inline examples/plot_replay_experiment.py:17]: just a placeholder as per our discussion (probably it's easier to add this as -optional?- dependence). 
Such a poor naming though :)

File: examples/plot_replay_experiment.py (hunk 2 above)
Comment (AlexandreAbraham) [reply examples/plot_replay_experiment.py:92]: Yes for now it's not clear because:
- I have this "GrowingIndex" (terrible name too) that is convenient to perform AL experiments
- I have variables that allow to track values and resume experiments.

My problem here is that I want to have a mix of both. One option is to modify variable to handle more complex types, such as this index. The other option is to allow using a tracked varaible inside the index. I have not found the best option yet.

File: examples/plot_replay_experiment.py (hunk 3 above)
Comment (AlexandreAbraham) [reply examples/plot_replay_experiment.py:97]: Yes, it's a setter

File: examples/plot_replay_cache.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+        config = dict(sampler=sampler_name)
+
+        with ReplayCache('./cache', './cache.db', keys=config) as cache:
+
+            index = GrowingIndex(X_train.shape[0])
+
+            # Add at least one sample from each class
+            index.add_to_selected([np.where(y_train == i)[0][0] for i in np.unique(y)])
+
+            selected = cache.variable('selected', index.selected)
```
[the remark this answers is not in the document]
Comment (AlexandreAbraham) [reply examples/plot_replay_cache.py]: Yes. Would persisted_variable be better?

File: examples/plot_replay_cache.py (hunk 1 above)
Comment (AlexandreAbraham) [reply examples/plot_replay_cache.py]: The problem is that the interest of the replay cache is to be able to compute a metric afterward. I could split it in two examples.

File: examples/plot_replay_cache.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+# Note the use of the GrowingIndex utils that facilitate the handing of
+# indices in an active learning experiment.
+
+def test_run(with_error=False, compute_metric=None):
+
+    for sampler_name, sampler in samplers:
+
+        config = dict(sampler=sampler_name)
+
+        with ReplayCache('./cache', './cache.db', keys=config) as cache:
```
[the remark this answers is not in the document]
Comment (AlexandreAbraham) [reply examples/plot_replay_cache.py]: Done

File: examples/plot_replay_cache.py [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+            index = GrowingIndex(X_train.shape[0])
+
+            # Add at least one sample from each class
+            index.add_to_selected([np.where(y_train == i)[0][0] for i in np.unique(y)])
```
[the remark this answers is not in the document]
Comment (AlexandreAbraham) [reply examples/plot_replay_cache.py]: Replaced by SampleSelector, and with docstring for a better understanding

File: examples/plot_replay_cache.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (AlexandreAbraham) [reply examples/plot_replay_cache.py]: Opted for persisted_value

File: examples/plot_replay_cache.py [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+
+            selected = cache.variable('selected', index.selected)
+
+            for j, prev_selected in cache.iter(range(n_iter), selected.previous()):
+                print('Computing iteration {}'.format(j))
+                index.resume(prev_selected)
```
[the remark this answers is not in the document]
Comment (AlexandreAbraham) [reply examples/plot_replay_cache.py]: I reworked the whole think to be clearer

File: examples/plot_replay_cache.py (hunk 1 above)
Comment (AlexandreAbraham) [reply examples/plot_replay_cache.py]: Split in 2 examples

File: examples/plot_replay_cache.py [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+# the resuming capability.
+#
+# Note the use of the GrowingIndex utils that facilitate the handing of
+# indices in an active learning experiment.
+
+def test_run(with_error=False, compute_metric=None):
+
+    for sampler_name, sampler in samplers:
+
+        config = dict(sampler=sampler_name)
```
[the remark this answers is not in the document]
Comment (AlexandreAbraham) [reply examples/plot_replay_cache.py]: Renamed `experiment_config`

File: examples/plot_replay_experiment.py [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+with ReplayCache(CACHE_PATH, DATABASE_PATH, keys=experiment_config) as cache:
+
+    # Create a selector with one sample from each class and persist it
+    init_selector = SampleSelector(X_train.shape[0])
+    init_selector.add_to_selected([np.where(y_train == i)[0][0] for i in np.unique(y)])
+    selector = cache.persisted_value('selector', init_selector)
+            
+    predictions = cache.persisted_value('prediction', None)
+
+    for j, prev_selected, prev_predictions in cache.iter(range(n_iter), selected.previous(), predictions.previous()):
```
Comment (dsleo) [inline examples/plot_replay_experiment.py:80]: `selected` doesn't seem tight, I guess it should be `selector` ?

File: examples/plot_replay_experiment.py [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+        print('Computing iteration {}'.format(j))
+        index.resume(prev_selected)
```
Comment (dsleo) [inline examples/plot_replay_experiment.py:82]: this `index`doesn't seem right either ? Still `selector` ?

File: examples/plot_replay_experiment.py (hunk 10 above)
Comment (AlexandreAbraham) [reply examples/plot_replay_experiment.py]: Fixed

File: examples/plot_replay_experiment.py (hunk 11 above)
Comment (AlexandreAbraham) [reply examples/plot_replay_experiment.py]: Fixed

File: examples/plot_replay_experiment.py (hunk 4 above)
Comment (dsleo) [reply examples/plot_replay_experiment.py:15]: ping :)

Comment (AlexandreAbraham) [reply examples/plot_replay_experiment.py:15]: Yay!

**Comment by AlexandreAbraham** [comment]
Additional note: using shelve can lead to a bad behavior. If the value store is not closed at the end of the script, segfault happens. See https://bugs.python.org/issue39672

File: examples/plot_replay_cache.py [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+
+samplers = [
+    ('margin', MarginSampler(model, batch_size))
+]
+
+#############################################################################
+# This function runs an experiment and optionally generate an error to test
+# the resuming capability.
+#
+# Note the use of the GrowingIndex utils that facilitate the handing of
```
[the remark this answers is not in the document]
Comment (AlexandreAbraham) [reply examples/plot_replay_cache.py]: I have completely changed the system.

**Comment by AlexandreAbraham** [comment]
Merging in order to move on. Please open issues if you want to add new comments
